File: python/336_PalindromePairs.py
from unittest import TestCase
# https://leetcode.com/problems/palindrome-pairs/


class PalindromePairs(TestCase):
    def palindromePairs(self, words):
        """
        :type words: List[str]
        :rtype: List[List[int]]
        """
        words_index = {}
        for (i, word) in enumerate(words):
            words_index[word] = i

        def isPal(word):
            for i in range(int(len(word)/2)):
                if word[i] != word[-1-i]:
                    return False
            return True
        ret = []
        for (i, word) in enumerate(words):
            for k in range(len(word)+1):
                if isPal(word[k:]):
                    j = words_index.get(word[:k][::-1])
                    if j not in (None, i):
                        ret.append([i, j])

                if k > 0 and isPal(word[:k]):
                    j = words_index.get(word[k:][::-1])
                    if j not in (None, i):
                        ret.append((j, i))

        return list(map(list, ret))

    def test1(self):
        self.assertEqual(sorted([[0,1],[1,0],[3,2],[2,4]]), sorted(self.palindromePairs(["abcd","dcba","lls","s","sssll"])))

    def test2(self):
        self.assertEqual(sorted([[0,1],[1,0]]), sorted(self.palindromePairs(["bat","tab","cat"])))

    def test3(self):
        self.assertEqual(sorted([[3,0],[1,3],[4,0],[2,4],[5,0],[0,5]]), sorted(self.palindromePairs(["a","b","c","ab","ac","aa"])))

File: kotlin/src/main/kotlin/336.PalindromePairs.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/palindrome-pairs/


class PalindromePairs{
  fun palindromePairs(words: Array<String>): List<List<Int>> {
    val ind = HashMap<String, Int>()
    for((i, w) in words.withIndex())
      ind[w] = i

    fun isPal(word: String) = word.reversed() == word

    val ret = mutableListOf<List<Int>>()
    for((i, word) in words.withIndex()){
      for(k in 0 until word.length+1){
        if(isPal(word.substring(k))) {
          val j = ind.getOrDefault(word.substring(0, k).reversed(), -1)
          if(j >= 0 && j != i)
            ret.add(listOf(i, j))
        }
        if(k > 0 && isPal(word.substring(0, k))){
          val j = ind.getOrDefault(word.substring(k).reversed(), -1)
          if(j >= 0 && j != i)
            ret.add(listOf(j, i))
        }
      }
    }
    return ret
  }

  @Test
  fun test1(){
    assertEquals(listOf(
        listOf(0,1),
        listOf(1,0),
        listOf(3,2),
        listOf(2,4)
    ), palindromePairs(arrayOf("abcd","dcba","lls","s","sssll")))
  }

  @Test
  fun test2(){
    assertEquals(listOf(
        listOf(0,1),
        listOf(1,0)
    ), palindromePairs(arrayOf("bat","tab","cat")))
  }

  @Test
  fun test3(){
    assertEquals(listOf(
        listOf(3,0),
        listOf(1,3),
        listOf(4,0),
        listOf(2,4),
        listOf(5,0),
        listOf(0,5)
    ), palindromePairs(arrayOf("a","b","c","ab","ac","aa")))
  }
}